lib/src/front_end/ast_node_visitor.dart

```diff
@@ -42,6 +42,10 @@ class AstNodeVisitor extends ThrowingAstVisitor<Piece> with PieceFactory {
   @override
   final CommentWriter comments;
 
+  /// The context set by the surrounding AstNode when visiting a child, or
+  /// [NodeContext.none] if the parent node doesn't set a context.
+  NodeContext _parentContext = NodeContext.none;
```

Comment:
This has a general name and description, but the current usage is pretty specific. Should we be worried that it might not generalize in practice?

Reply:
I'm somewhat worried.

There are an increasing number of places where we need to look at both parent and child to decide how to format something, and there are a number of different strategies we take to do that. I would really like something more consistent and unified, but I haven't figure out what yet. I have some aspirations that this could be used for more things, but haven't taken the time to refactor other code to using this yet.

Or maybe there's some other better solution out there that I haven't found yet. One idea in the back of my head is that maybe Pieces should encode a little more about the original syntax nodes they are based on? Right now, if you want to choose how to format a parent based on its children, you need to make that choice *before* the children are converted to Pieces, because Pieces tend to drop too much information on the floor. (Dropping stuff on the floor is sort of the *point* because it lets different kinds of syntax share the same formatting and means we need far fewer Piece classes than AST node classes. But it makes situations like this harder.)

I'm not sure.

Reply:
It feels brittle to me - you can't know if `_parentContext` is an _immediate_ parent, or if there is some chain of ancestry that didn't care to set that field.

If being able to preserve context for a non-immediate ancestor is a feature - then we need to worry about any ways where nodes using this strategy could interleave in the ancestry chain with the nodes they are trying to communicate context to.

Reply:
It will always be the context of your immediate parent. More concretely, it will be whatever context that the parent node passes in when it calls `nodePiece()` (which every way of visiting a node ultimately bottoms out on) or it will be `NodeContext.none`. If you don't pass in a context when calling `nodePiece()`, it defaults to `NodeContext.none`. It doesn't inherit the surrounding one.